pkg/cyclic/mark.go

```diff
@@ -96,98 +95,56 @@ func (m MarkMap) shouldFilterTxn(startTs uint64, filterReplicaIDs []uint64) (*mo
+							zap.Uint64("start-ts", event.StartTs),
+							zap.Any("first", first),
+							zap.Any("second", event))
 					}
-					markMap[event.StartTs] = event
-				}
-				continue
-			}
-			for _, event := range events {
-				sameTxn, txnFound := output[event.StartTs]
-				if !txnFound {
-					sameTxn = make(map[model.TableName][]*model.RowChangedEvent)
-				}
-				table, tableFound := sameTxn[name]
-				if !tableFound {
-					table = make([]*model.RowChangedEvent, 0, 1)
 				}
-				table = append(table, event)
-				sameTxn[name] = table
-				output[event.StartTs] = sameTxn
+				markMap[event.StartTs] = event
 			}
 		}
 	}
-	return
-}
-
-// ReduceCyclicRowsGroup filters duplicate rows bases on filterReplicaIDs and
-// join mark table rows and normal table rows into one group if start ts is
-// the same.
-func ReduceCyclicRowsGroup(
-	input TxnMap, markMap MarkMap, filterReplicaIDs []uint64,
-) map[model.TableName][][]*model.RowChangedEvent {
-	output := make(map[model.TableName][][]*model.RowChangedEvent, len(input))
-
-	for startTs, txn := range input {
-		// Check if we should skip this event
-		markRow, needSkip := markMap.shouldFilterTxn(startTs, filterReplicaIDs)
-		if needSkip {
-			// Found cyclic mark, skip this event as it originly created from
-			// downstream.
+	for table, txns := range txnsMap {
+		if IsMarkTable(table.Schema, table.Table) {
+			delete(txnsMap, table)
 			continue
 		}
-
-		for name, events := range txn {
-			if startTs != events[0].StartTs {
-				panic(fmt.Sprintf(
-					"start ts mismatch %d != %d", startTs, events[0].StartTs))
+		filteredTxns := make([]*model.Txn, 0, len(txns))
+		for _, txn := range txns {
+			// Check if we should skip this event
+			markRow, needSkip := markMap.shouldFilterTxn(txn.StartTs, filterReplicaIDs)
+			if needSkip {
+				// Found cyclic mark, skip this event as it originly created from
+				// downstream.
+				continue
 			}
-			multiRows := make([][]*model.RowChangedEvent, 0, 1)
-			rows := make([]*model.RowChangedEvent, 0, len(events)+1)
+			txn.ReplicaID = replicaID
 			if markRow != nil {
-				var mark model.RowChangedEvent = *markRow
-				// Rewrite mark table name based on event's table ID.
-				schema, table := MarkTableName(name.Schema, name.Table)
-				mark.Table = &model.TableName{
-					Schema: schema,
-					Table:  table,
-				}
-				rows = append(rows, &mark)
+				txn.ReplicaID = ExtractReplicaID(markRow)
 			}
-			rows = append(rows, events...)
-
-			if rs, ok := output[name]; ok {
-				multiRows = append(rs, rows)
-			} else {
-				multiRows = append(multiRows, rows)
-			}
-			output[name] = multiRows
+			filteredTxns = append(filteredTxns, txn)
+		}
+		if len(filteredTxns) == 0 {
+			delete(txnsMap, table)
+		} else {
+			txnsMap[table] = filteredTxns
 		}
 	}
-	for _, events := range output {
-		// Per table order may lose during map stage, we need to sort events.
-		sort.Slice(events, func(i, j int) bool { return events[i][0].CommitTs < events[j][0].CommitTs })
```

Comment:
Why can it be removed?

Reply:
I didn't add `mark table event` into `events`, so we didn't lose the order in my implement